datafusion/catalog/src/session.rs

```diff
@@ -0,0 +1,102 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use async_trait::async_trait;
+use datafusion_common::config::ConfigOptions;
+use datafusion_common::{DFSchema, Result};
+use datafusion_execution::config::SessionConfig;
+use datafusion_execution::runtime_env::RuntimeEnv;
+use datafusion_execution::TaskContext;
+use datafusion_expr::execution_props::ExecutionProps;
+use datafusion_expr::{AggregateUDF, Expr, LogicalPlan, ScalarUDF, WindowUDF};
+use datafusion_physical_plan::{ExecutionPlan, PhysicalExpr};
+use std::any::Any;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+#[async_trait]
+pub trait CatalogSession: Send + Sync {
```

Comment:
I think prior to merging this PR we should document this trait a bit (I can help, but likely not until tomorrow)

I think it will also help people to upgrade if we add an example of how to get an actual `SessionState` reference from it (as a way to ease the migration)

```rust
let session_context = catalog_session.
  .as_any()
  .downcast_ref::<SessionState>()
  .unwrap();
```

Reply:
@alamb that would help a lot. In addition this comment should be clarified on the `SessionState` struct:

```
/// Get a new TaskContext to run in this session
    pub fn task_ctx(&self) -> Arc<TaskContext> {
        Arc::new(TaskContext::from(self))
    }
```
What I would want is to get the `TaskContext` for a specific query that was initialized by custom code, I think this will create a new one (using the `SessionState` config). My question is: how to get the `TaskContext` passed into a fn like `supports_filters_pushdown`? I am hoping with this release that this will be feasible.

Reply:
Added a new commit with some docs.
Please help me improve them, it doesn't feel complete yet.


> I think it will also help people to upgrade if we add an example of how to get an actual `SessionState` reference from it (as a way to ease the migration)
> 
> ```rust
> let session_context = catalog_session.
>   .as_any()
>   .downcast_ref::<SessionState>()
>   .unwrap();
> ```

i like the idea.
I added this as preformatted text, but it should be a code snippet that `cargo doc` can check for compilation.
How should i handle this without introducing circular dependency between the new catalog module and the core?

Reply:
I have some thoughts -- let me give it a try.

Reply:
Here is a proposed PR to this branch to improve the comments: https://github.com/findepi/datafusion/pull/4 with a doc example

Reply:
thank you, that's well written!